SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift

```diff
@@ -0,0 +1,214 @@
+
+  let substitutionMap = context.getContextSubstitutionMap(for: type)
+  builder.createApply(function: functionRef, substitutionMap, arguments: [object])
+  context.erase(instruction: release)
+}
+
+private func stripRCIdentityPreservingInsts(_ value: Value) -> Value? {
+  switch value {
+  // First strip off RC identity preserving casts.
+  case let inst as Instruction where inst is UpcastInst ||
+    inst is UncheckedRefCastInst ||
+    inst is InitExistentialRefInst ||
+    inst is OpenExistentialRefInst ||
+    inst is RefToBridgeObjectInst ||
+    inst is BridgeObjectToRefInst ||
+    inst is ConvertFunctionInst ||
+    inst is UncheckedEnumDataInst:
+    return inst.operands[0].value
+
+  // Then if we have a struct_extract that is extracting a non-trivial member
+  // from a struct with no other non-trivial members, a ref count operation on
+  // the struct is equivalent to a ref count operation on the extracted
+  // member. Strip off the extract.
+  case let sei as StructExtractInst where sei.isFieldOnlyNonTrivialField:
+    return sei.operands[0].value
+
+  // If we have a struct instruction with only one non-trivial stored field, the
+  // only reference count that can be modified is the non-trivial field. Return
+  // the non-trivial field.
+  case let si as StructInst:
+    return si.uniqueNonTrivialOperand
+
+  // If we have an enum instruction with a payload, strip off the enum to
+  // expose the enum's payload.
+  case let ei as EnumInst where !ei.operands.isEmpty:
+    return ei.operands[0].value
+
+  // If we have a tuple_extract that is extracting the only non trivial member
+  // of a tuple, a retain_value on the tuple is equivalent to a retain_value on
+  // the extracted value.
+  case let tei as TupleExtractInst where tei.isEltOnlyNonTrivialElt:
+    return tei.operands[0].value
+
+  // If we are forming a tuple and the tuple only has one element with reference
+  // semantics, a retain_value on the tuple is equivalent to a retain value on
+  // the tuple operand.
+  case let ti as TupleInst:
+    return ti.uniqueNonTrivialOperand
+
+  default:
+    return nil
+  }
+}
+
+private extension Instruction {
+  /// Search the operands of this tuple for a unique non-trivial elt. If we find
+  /// it, return it. Otherwise return `nil`.
+  var uniqueNonTrivialOperand: Value? {
+    var candidateElt: Value?
+    let function = self.function
+
+    // For each operand...
+    for op in operands.enumerated() {
+      // If the operand is not trivial...
+      if !op.type.isTrivial(in: function) {
+        // And we have not found a `candidateElt` yet, set index to `op` and continue.
+        if candidateElt == nil {
+          candidateElt = op
+          continue
+        }
+
+        // Otherwise, we have two values that are non-trivial. Bail.
+        return nil
+      }
+    }
+
+    return candidateElt
+  }
+}
+
+private extension TupleExtractInst {
+  var isEltOnlyNonTrivialElt: Bool {
```

Comment:
For the use in the release-devirtualizer it would be simpler to add a `Type.countNonTrivialTupleElements` and bail if it's not 1.